### DISCUSSION TITLE
Real time object based rendering

### DISCUSSION THREAD

**Discussion by vertexgamer** [VoidXH/Cavern#185] [Ideas]
Hello, i was looking for a solution to play object based audio in realtime, unfortunately i noticed cavern and cavernize only supports playing or converting files.
Would it be possible to make cavern install a virtual audio device with atmos capabilities (which the user will select as the default output) and then letting cavern handle the conversion and mapping of object based audio to real physical hardware?

**Comment by VoidXH** [comment]
Many methods were tried, but it was always too large of a task. A complete media player is in the works that will have Atmos decoding and rerendering capabilities, but it's very far away.

**Reply by EaryChow** [reply]
Is the said media player going to include a virtual audio device to recieve signals from, say, browser DD+JOC stream? Or still local files only? If it's still local file only it's still very limited.

**Reply by VoidXH** [reply]
Local files only. Browser streams are currently not readable without making an entire audio driver.

**Reply by EaryChow** [reply]
There is one way I can think of that might be doable.

Chromium browsers (specifically talking about Edge and Thorium browser, Google Chrome doesn't seem to support EAC3 stream) can enable "Media Foundation for Clear" in the browser feature flags, to enable bitstream/passthrough (Actually not sure about what Media Foundation for Clear actually does, my current understanding is bitstreaming, but I might be wrong). Then we can use a free software called Voicemeeter, which uses a virtual speaker device for the browser audio to send to, and it bitstreams the signals to its virtual microphone device. Then the potential real time Cavern media player can probably take signals from this virtual microphone and live decode the Atmos content, which would then be sent to the actual playback device.

Using this procedure, the media player only needs to be able to take input from a virtual microphone device and live decode it, and output to a user-selected device, other stuff can be left to software like Voicemeeter to handle.

Just an idea, not sure about whether the technical details would work out, just some thoughts on this that I hope might be helpful.

**Reply by Firerouge** [reply]
Do you know if there's any existing public documentation on using [voicemeeter virtual audio device](https://vb-audio.com/Cable/index.htm) with the [CavernizeRealtime API](https://cavern.sbence.hu/cavern/dev/api/Cavern/CavernizeRealtime/index.php) to decode atmos bitstreams in realtime?

**Reply by VoidXH** [reply]
The CavernizeRealtime API is an upconverter, not an Atmos render. It's available as Cavarnize VST, and runs in real time with Equalizer APO. True Atmos decoding is not available in real time.

**Reply by KernelLand32** [reply]
Any progress on this so far?

**Reply by VoidXH** [reply]
> Any progress on this so far?

See #323 and #201.

**Reply by KernelLand32** [reply]
Interesting. Would also be a wonderful idea to integrate with [this HRTF project](https://github.com/ShanonPearce/ASH-Toolset), don't know if you've seen it previously. To get a more neutral reference point and avoid the traditional phasey/bathroom reverb sound. Then you'd be able to decode Atmos in real-time via `mpv`
